IPC::Semaphore waits don't terminate if the remote process crashes and can cause hangs.

mattwoodrow opened the pull request:
#### f2796edab1163daa5bb37135f7fee0dbe58f577e
<pre>
IPC::Semaphore waits don&apos;t terminate if the remote process crashes and can cause hangs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=260690">https://bugs.webkit.org/show_bug.cgi?id=260690</a>
&lt;rdar://114591255&gt;

Reviewed by Kimmo Kinnunen.

This introduces new IPC::Event and Signal classes (implemented properly for cocoa
only so far) that can detect crashes (by the &apos;Signal&apos; instance being destroyed).
It reimplements a semaphore using a mach_port_t, and uses no sender notifications to
detect when the last MachSendRight (owned by the Signal component) is released.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::setFlushSignal):
(WebKit::RemoteDisplayListRecorder::flushContext):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/IPCEvent.h: Added.
(IPC::Signal::signal):
(IPC::Signal::encode):
(IPC::Signal::decode):
(IPC::Signal::Signal):
(IPC::Event::Event):
(IPC::Event::operator=):
(IPC::Event::wait):
(IPC::Event::waitFor):
(IPC::createEventSignalPair):
* Source/WebKit/Platform/IPC/IPCSemaphore.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
* Source/WebKit/Platform/IPC/darwin/IPCEventDarwin.cpp: Added.
(IPC::requestNoSenderNotifications):
(IPC::clearNoSenderNotifications):
(IPC::Signal::signal):
(IPC::Signal::encode):
(IPC::Signal::decode):
(IPC::createEventSignalPair):
(IPC::Event::~Event):
(IPC::Event::wait):
(IPC::Event::waitFor):
* Source/WebKit/Platform/IPC/darwin/IPCSemaphoreDarwin.cpp:
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(types_that_cannot_be_forward_declared):
(argument_coder_headers_for_type):
(headers_for_type):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::setFlushSignal):
(WebKit::RemoteDisplayListRecorderProxy::flushContext):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxyFlushFence::create):
(WebKit::RemoteImageBufferProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferProxyFlushFence::tryTakeEvent):
(WebKit::RemoteImageBufferProxyFlushFence::RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxy::flushDrawingContextAsync):
(WebKit::RemoteImageBufferProxyFlushFence::tryTakeSemaphore): Deleted.

Canonical link: <a href="https://commits.webkit.org/268093@main">https://commits.webkit.org/268093@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08d575b669e7923cf875738655188f325fd1862b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18604 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19545 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20465 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17410 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18800 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19260 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18987 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16192 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21341 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/18771 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16209 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16954 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23404 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17230 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17124 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21299 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17726 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15025 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16779 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4436 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21147 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple reviewed (commented):
Sorry, didn't get to this, started too late.
One inconvenient thing I was thinking was that maybe the CrashAwareSemaphore isn't such a good name. I suppose semaphores count typically? 
Maybe something like  IPC::[something?]Event and IPC::[something?]Signal  or something?

kkinnunen-apple commented on Source/WebKit/Platform/IPC/darwin/IPCSemaphoreDarwin.cpp:
```diff
@@ -121,4 +150,103 @@ void Semaphore::destroy()
+    memset(&receiveMessage, 0, sizeof(receiveMessage));
+    mach_msg_return_t ret = mach_msg(&receiveMessage.header, MACH_RCV_MSG, 0, sizeof(receiveMessage), m_receiveRight, MACH_MSG_TIMEOUT_NONE, MACH_PORT_NULL);
+    if (ret != MACH_MSG_SUCCESS) {
+        WTFLogAlways("recv message failed");
```
You could replace the wtflog with something or remove.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/darwin/IPCSemaphoreDarwin.cpp:
```diff
@@ -121,4 +150,103 @@ void Semaphore::destroy()
+        return false;
+    }
+    if (receiveMessage.header.msgh_id != inlineBodyMessageID) {
+        WTFLogAlways("recv message returned other message type");
```
You could replace the wtflog with something or remove.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/IPCSemaphore.h:
```diff
@@ -85,4 +94,93 @@ class Semaphore {
 };
 
+// FIXME: Write proper interruptible implementations for non-COCOA platforms.
+class InterruptibleBinarySemaphoreSignaler {
```
I'd still prefer IPC::Event in (IPCEvent.h) on the count of this being quite a mouthful.
With the documentation that it's interruptible.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/IPCSemaphore.h:
```diff
@@ -85,4 +94,93 @@ class Semaphore {
+#endif
+};
+
+class InterruptibleBinarySemaphoreWaiter {
```
I'd still prefer IPC::Signal (in IPCEvent.h or IPCSignal.h).

kkinnunen-apple commented on Source/WebKit/Platform/IPC/IPCSemaphore.h:
```diff
@@ -85,4 +94,93 @@ class Semaphore {
+
+#if PLATFORM(COCOA)
+    Lock m_lock;
+    mach_port_t m_receiveRight;
```
WTF_GUARDED_BY(m_lock); perhaps?

kkinnunen-apple replied:
In fact, the lock gives the impression that this would be thread-safe, but I'm not sure that makes sense.
For example, the move operation is not.
Contrast with IPC::Semaphore, there's no lock there.

kkinnunen-apple commented on Source/WebKit/Platform/IPC/IPCSemaphore.h:
```diff
@@ -85,4 +94,93 @@ class Semaphore {
+#endif
+};
+
+struct InterruptibleBinarySemaphore {
```
Appears `struct InterruptibleBinarySemaphore` is quite useless, as InterruptibleBinarySemaphore::Pair is the payload struct.

I'd prefer:
```
struct EventSignalPair {
 ...
};
std::optional<EventSignalPair> createEventSignalPair();
```

kkinnunen-apple reviewed (commented):
Could still benefit from some sanity check use tests in TestIPC.

kkinnunen-apple reviewed (approved):
Nice!

kkinnunen-apple commented on Source/WebKit/Platform/IPC/IPCSemaphore.h:
```diff
@@ -40,9 +41,19 @@
 
 namespace IPC {
 
+struct InterruptibleBinarySemaphore;
```
Here still the ole name

mattwoodrow committed:
IPC::Semaphore waits don't terminate if the remote process crashes and can cause hangs.
https://bugs.webkit.org/show_bug.cgi?id=260690
<rdar://114591255>

Reviewed by Kimmo Kinnunen.

This introduces new IPC::Event and Signal classes (implemented properly for cocoa
only so far) that can detect crashes (by the 'Signal' instance being destroyed).
It reimplements a semaphore using a mach_port_t, and uses no sender notifications to
detect when the last MachSendRight (owned by the Signal component) is released.

* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::setFlushSignal):
(WebKit::RemoteDisplayListRecorder::flushContext):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/Platform/IPC/IPCEvent.h: Added.
(IPC::Signal::signal):
(IPC::Signal::encode):
(IPC::Signal::decode):
(IPC::Signal::Signal):
(IPC::Event::Event):
(IPC::Event::operator=):
(IPC::Event::wait):
(IPC::Event::waitFor):
(IPC::createEventSignalPair):
* Source/WebKit/Platform/IPC/IPCSemaphore.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
* Source/WebKit/Platform/IPC/darwin/IPCEventDarwin.cpp: Added.
(IPC::requestNoSenderNotifications):
(IPC::clearNoSenderNotifications):
(IPC::Signal::signal):
(IPC::Signal::encode):
(IPC::Signal::decode):
(IPC::createEventSignalPair):
(IPC::Event::~Event):
(IPC::Event::wait):
(IPC::Event::waitFor):
* Source/WebKit/Platform/IPC/darwin/IPCSemaphoreDarwin.cpp:
* Source/WebKit/Platform/SourcesCocoa.txt:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
(types_that_cannot_be_forward_declared):
(argument_coder_headers_for_type):
(headers_for_type):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::setFlushSignal):
(WebKit::RemoteDisplayListRecorderProxy::flushContext):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp:
(WebKit::RemoteImageBufferProxyFlushFence::create):
(WebKit::RemoteImageBufferProxyFlushFence::waitFor):
(WebKit::RemoteImageBufferProxyFlushFence::tryTakeEvent):
(WebKit::RemoteImageBufferProxyFlushFence::RemoteImageBufferProxyFlushFence):
(WebKit::RemoteImageBufferProxy::flushDrawingContextAsync):
(WebKit::RemoteImageBufferProxyFlushFence::tryTakeSemaphore): Deleted.

Canonical link: https://commits.webkit.org/268093@main
